sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala

```diff
@@ -192,7 +192,14 @@ private[sql] case class PreWriteCheck(catalog: Catalog) extends (LogicalPlan =>
         }
 
         PartitioningUtils.validatePartitionColumnDataTypes(
-          query.schema, partitionColumns, catalog.conf.caseSensitiveAnalysis)
+          c.child.schema, c.partitionColumns, catalog.conf.caseSensitiveAnalysis)
+
+        c.bucketSpec.foreach(_.sortingColumns.foreach(_.foreach { sortCol =>
```

Comment:
these 3 foreaches nested in one line is too crazy. Let's simplify this (even if it means more loc)